control_plane/src/attachment_service.rs

```diff
@@ -220,9 +241,213 @@ impl AttachmentService {
             .expect("non-Unicode path")
     }
 
+    /// PIDFile for the postgres instance used to store attachment service state
+    fn postgres_pid_file(&self) -> Utf8PathBuf {
+        Utf8PathBuf::from_path_buf(
+            self.env
+                .base_data_dir
+                .join("attachment_service_postgres.pid"),
+        )
+        .expect("non-Unicode path")
+    }
+
+    /// In order to access database migrations, we need to find the Neon source tree
+    async fn find_source_root(&self) -> anyhow::Result<Utf8PathBuf> {
+        // We assume that either prd or our binary is in the source tree. The former is usually
+        // true for automated test runners, the latter is usually true for developer workstations. Often
+        // both are true, which is fine.
+        let candidate_start_points: Vec<Utf8PathBuf> = vec![
+            // Current working directory
+            Utf8PathBuf::from_path_buf(std::env::current_dir()?).unwrap(),
+            // Directory containing the binary we're running inside
+            Utf8PathBuf::from_path_buf(env::current_exe()?.parent().unwrap().to_owned()).unwrap(),
+            // Source dir on github action workers
+            Utf8PathBuf::from_str("/__w/neon").unwrap(),
+        ];
+
+        // For each candidate start point, search through ancestors looking for a neon.git source tree root
+        for start_point in &candidate_start_points {
+            // Start from the build dir: assumes we are running out of a built neon source tree
+            let mut cursor = start_point.clone();
+            loop {
```

Comment:
```suggestion
            for path in start_point.ancestors() {
```

Reply:
Applied by hand.  I wouldn't use .ancestors() in the main code because it's sync, but it's fine here.